src/Blocks/BlockButton.h

```diff
@@ -133,12 +123,119 @@ class cBlockButtonHandler :
 	}
 
 	/** Extracts the ON bit from metadata and returns if true if it is set */
-	static bool IsButtonOn(NIBBLETYPE a_BlockMeta)
+	static bool IsButtonOn(NIBBLETYPE a_Meta)
 	{
-		return ((a_BlockMeta & 0x8) == 0x8);
+		return (a_Meta & 0x08) == 0x08;
 	}
-} ;
 
+	/** Event handler for an arrow striking a block.
+	Performs appropriate handling if the arrow intersected a wooden button. */
+	static void OnArrowHit(cWorld & a_World, const Vector3i a_Position, const eBlockFace a_HitFace)
+	{
+		BLOCKTYPE Type;
+		NIBBLETYPE Meta;
+		const auto Pos = AddFaceDirection(a_Position, a_HitFace);
+
+		if (
+			!a_World.GetBlockTypeMeta(Pos, Type, Meta) ||
+			IsButtonOn(Meta) ||
+			!IsButtonPressedByArrow(a_World, Pos, Type, Meta)
+		)
+		{
+			// Bail if we're not specifically a wooden button, or it's already on
+			// or if the arrow didn't intersect. It is very important that nothing is
+			// done if the button is depressed, since the release task will already be queued
+			return;
+		}
+
+		a_World.SetBlockMeta(Pos, Meta | 0x08, false);
+		a_World.WakeUpSimulators(Pos);
+
+		// sound name is ok to be wood, because only wood gets triggered by arrow
+		a_World.GetBroadcastManager().BroadcastSoundEffect("block.wood_button.click_on", Pos, 0.5f, 0.6f);
+
+		// Queue a button reset
+		QueueButtonRelease(a_World, Pos, Type);
+	}
+
+private:
+
+	/** Schedules a recurring event at appropriate intervals to release a button at a given position.
+	The given block type is checked when the task is executed to ensure the position still contains a button. */
+	static void QueueButtonRelease(cWorld & a_Wurld, const Vector3i a_Position, const BLOCKTYPE a_BlockType)
+	{
+		const auto TickDelay = (a_BlockType == E_BLOCK_STONE_BUTTON) ? 20 : 30;
+		a_Wurld.ScheduleTask(
+			TickDelay,
+			[a_Position, a_BlockType](cWorld & a_World)
+			{
+				BLOCKTYPE Type;
+				NIBBLETYPE Meta;
+
+				if (
+					!a_World.GetBlockTypeMeta(a_Position, Type, Meta) ||
+					(Type != a_BlockType) || !IsButtonOn(Meta)
+				)
+				{
+					// Total failure or block changed, bail
+					return;
+				}
+
+				if (IsButtonPressedByArrow(a_World, a_Position, Type, Meta))
+				{
+					// Try again in a little while
+					QueueButtonRelease(a_World, a_Position, a_BlockType);
+					return;
+				}
```

Comment:
Instead of re-queuing, I think the arrows should should re-queue the release when they are destroyed (collected or despawned)

Reply:
I tried but couldn't resolve this issue:

Right click wooden button (causes a queue), immediately fire arrow into it, collect arrow (as you suggest, causes a queue). Now we have two queues; after the first fires, the second ends up fighting the player if they press the button again.

Reply:
The code currently resolves this by using the depressed state as a "queue in flight" indicator, so nothing adds a new queued task if the button is pressed, and the existing task keeps itself alive until it can release.

Reply:
So either we eat up CPU, or we get a bad very-corner-case. Oh well... Can you even simulate that corner case in vanilla? :P

Reply:
I'm half convinced vanilla does this based on how it behaves during testing...

Reply:
Why not just let the redstone handler deal with the timing issues? I don't see why the noop simulator needs to have functioning buttons. It's seems natural that redstone devices would behave differently under a different simulator.

Reply:
Unfortunately that would mean that with the Noop simulator, all buttons will end up forever-pressed eventually, from the players pressing them.
I think this solution is the best, the CPU price is very low.